File: original-challenges/Order_Up/cb_10/src/table.c
/*
 * Copyright (C) Narf Industries <info@narfindustries.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/
#include <libcgc.h>
#include <stdint.h>
#include "libc.h"
#include "memset.h"
#include "malloc.h"

#include "menu.h"
#include "table.h"

Table table;

/**
 * Calculate how many seats this table will have.
 *
 * @return Number of seats.
 */
static unsigned char calculate_seats_count(void) {
	unsigned char seats = 1;
	unsigned char *flag_bytes = (unsigned char *)FLAG_PAGE;
    // for testing
    // seats = 5;
    // for release
    seats = flag_bytes[table.id] % MAX_TABLE_SIZE;
    if (MIN_TABLE_SIZE > seats) {
       seats += MIN_TABLE_SIZE;
    }

	return seats;
}

/**
 * Count the number of occupied seats
 *
 * @return Number of occupied seats
 */
static unsigned char get_occupied_seats_count(void) {
	unsigned char occupied_seats = 0;
	Customer *cur = table.customers;
	while (NULL != cur) {
		cur = (Customer *)cur->next;
		occupied_seats++;
	}
	DBG("%U occupied seats\n", occupied_seats);
	return occupied_seats;
}

/**
 * Count the number of empty seats
 *
 * @return Number of empty seats.
 */
static unsigned char get_empty_seats_count(void) {
	unsigned char empty_seats = get_seats_count() - get_occupied_seats_count();
	DBG("%U empty seats\n", empty_seats);
	return empty_seats;
}

/**
 * Update the table status
 *
 * @param ts 	New table status
 */
static void set_table_status(TableStatus ts) {
	table.status = ts;
}

/**
 * Seat one customer
 *
 * @param c 	Customer to seat
 * @return SUCCESS on success, else -1 if seats are full
 */
static int seat_customer(Customer *c) {
	DBG("seating a customer\n");
	Customer *cur = table.customers;

	if (0 == get_empty_seats_count()) {
		DBG("Customer %U not seated. 0 empty seats\n", c->id);
		return -1;
	}

	c->next = NULL;
	if (NULL == cur) {
		table.customers = c;
	} else {
		for (; NULL != cur->next; cur=(Customer *)cur->next);

		cur->next = c;
	}

	DBG("Customer %U seated\n", c->id);
	return SUCCESS;
}

/**
 * Take the order of one customer
 *
 * @param c 	Customer
 * @return Pointer to a new Order or NULL if no order
 */
static Order *take_customer_order(Customer *c) {
	Order *o = malloc(sizeof(Order));
	MALLOC_OK(o);
	memset(o, '\0', sizeof(Order));

	o->t_id = table.id;
	o->c_id = c->id;
	DBG("Table status: %U\n", table.status);
	switch(table.status) {
		case APP_RTO:	// RTO => ready to order
			o->ftype = APP_TYPE;
			o->item = select_appetizer(get_appetizer_list(), c);
			DBG("Customer %U ordering appetizer\n", c->id);
			break;
		case MEAL_RTO:
			o->ftype = MEAL_TYPE;
			o->item = select_meal(get_meal_list(), c);
			DBG("Customer %U ordering meal\n", c->id);
			break;
		case DES_RTO:
			o->ftype = DES_TYPE;
			o->item = select_dessert(get_dessert_list(), c);
			DBG("Customer %U ordering dessert\n", c->id);
			break;
		default:		// WAIT => already ordered, VACANT, FINISHED
			DBG("Customer %U NOT ordering\n", c->id);
			free(o);
			return NULL;
	}
	return o;
}

void table_setup(void) {
	memset(&table, '\0', sizeof(Table));
	table.id = TABLE_ID;
	table.status = VACANT;
	table.seats = calculate_seats_count();
	table.customers = NULL;
}

void table_clean_and_set(void) {
	Customer *c_list = table.customers;
	Customer *c = NULL;
	unsigned char cnt = get_customer_list_size(c_list);
	DBG("%U Customers to remove from table\n", cnt);

	c = pop_customer_from_list(&c_list);
	while (NULL != c) {
		DBG("Customer %U left the table\n", c->id);
		free(c);
		c = pop_customer_from_list(&c_list);
	}

	table_setup();
}

unsigned char get_seats_count(void) {
	DBG("%U total seats\n", table.seats);
	return table.seats;
}

TableStatus get_table_status(void) {
	return table.status;
}

int seat_customers(unsigned char qty, Customer *c_list) {
	int ret = SUCCESS;
	Customer *c = NULL;

	if (VACANT != get_table_status()) {
		DBG("table is not vacant");
		clear_customer_list(&c_list);
		return -1;
	}

    DBG("seating %U customers\n", qty);

	for (unsigned char cnt = 0; cnt < qty; cnt++) {
		c = pop_customer_from_list(&c_list);
		if (NULL == c) return -1;
		ret = seat_customer(c);
		if (SUCCESS != ret) return ret;
	}

	if (0 < qty) {
		set_table_status(table.status + 1);
		DBG("table status updated: %U\n", table.status);
	} else {
		DBG("table status NOT updated: %U\n", table.status);
	}
	return SUCCESS;
}

unsigned char take_customer_orders(Order **o_list) {
	unsigned char o_count = 0;
	Customer *cur = table.customers;
	Order *o = NULL;

	while (NULL != cur) {
		o = take_customer_order(cur);
		if (NULL != o) {
			append_order_to_list(o_list, o);
			o_count++;
		}

		cur = (Customer *)cur->next;		
	}
	if (0 < o_count) {
		set_table_status(table.status + 1);
		DBG("table status updated: %U\n", table.status);
	} else {
		DBG("table status NOT updated: %U\n", table.status);
	}
	return o_count;
}

/**
 * When food is to be delivered, determine the type of food based on the TableStatus.
 *
 * @param ts 	TableStatus
 * @return Type of food if this table is ready for food delivery, else FOOD_TYPES
 */
static FoodTypes get_delivery_foodtype_from_tablestatus(TableStatus ts) {
	FoodTypes ftype = 0;

	switch(table.status) {
		case APP_WAIT:
			ftype = APP_TYPE;
			break;
		case MEAL_WAIT:
			ftype = MEAL_TYPE;
			break;
		case DES_WAIT:
			ftype = DES_TYPE;
			break;
		default:		// RTO => ready to order, VACANT, FINISHED
			ftype = FOOD_TYPES;
	}
	return ftype;
}

/**
 * Give a food item to a customer
 *
 * @param c 		Customer
 * @param ftype 	Type of food
 * @param food 		Food item
 * @return SUCCESS on success, else -1 if invalid ftype
 */
static char give_food_to_customer(Customer *c, FoodTypes ftype, void *food) {

	char ret = SUCCESS;
	unsigned int bufsz = 0;
	void *buf = NULL;

	// get size of food using provided ftype
	switch (ftype) {
		case APP_TYPE:
			DBG("food size as appetizer\n");
			bufsz = sizeof(Appetizer);
			break;
		case MEAL_TYPE:
			DBG("food size as Meal\n");
			bufsz = sizeof(Meal);
			break;
		case DES_TYPE:
			DBG("food size as Dessert\n");
			bufsz = sizeof(Dessert);
			break;
		default:
			DBG("invalid ftype\n");
			ret = -1;
	}

#ifndef PATCHED_9
	// VULN: table.status can be mismatched with ftype provided by user.
	// Use table.status to get location to write food into Customer and the proper food inspection function
	ftype = get_delivery_foodtype_from_tablestatus(table.status);
#endif
	switch (ftype) {
		case APP_TYPE:
			DBG("buf as appetizer\n");
			buf = &c->a;
			c->fi = inspect_appetizer;
			break;
		case MEAL_TYPE:
			DBG("buf as Meal\n");
			buf = &c->m;
			c->fi = inspect_meal;
			break;
		case DES_TYPE:
			DBG("buf as Dessert\n");
			buf = &c->d;
			c->fi = inspect_dessert;
			break;
		default:
			DBG("invalid ftype\n");
			ret = -1;
	}

	if (SUCCESS == ret) {
		accept_food(food, buf, bufsz);
		ret = c->fi(buf, c->id);
	}

#ifdef DEBUG
	switch (ftype) {
		case APP_TYPE:
			DBG("wrote appetizer\n");
			print_appetizer(&c->a);
			break;
		case MEAL_TYPE:
			DBG("wrote meal\n");
			print_meal(&c->m);
			break;
		case DES_TYPE:
			DBG("wrote dessert\n");
			print_dessert(&c->d);
			break;
		default:
			DBG("invalid ftype\n");
			ret = -1;
	}
	print_customer(c);
#endif

	return ret;
}

unsigned char deliver_customer_orders(Order **o_list) {
	unsigned char o_count = get_order_list_size(*o_list);
	unsigned char d_count = 0;
	Order *o = NULL;
	Customer *c = NULL;
	FoodTypes ft = 0;

	// loop through orders
	o = pop_order_from_list(o_list);
	while (NULL != o) {

		// get customer having c_id from order
		c = get_customer_by_id(table.customers, o->c_id);
		if (NULL != c) {
			// write food to customer
			if (SUCCESS == give_food_to_customer(c, o->ftype, o->item)) {
				d_count++;
			}
		}
		free(o->item);
		free(o);
		o = pop_order_from_list(o_list);
	}

	if (0 < d_count) {
		set_table_status(table.status + 1);
		DBG("table status updated: %U\n", table.status);
	} else {
		DBG("table status NOT updated: %U\n", table.status);
	}

	return (o_count - d_count);
}

File: original-challenges/Order_Up/support/table.py
#!/usr/bin/env python
#
# Copyright (C) 2014 Narf Industries <info@narfindustries.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
from random import randint

import support as sp
from common import DEBUG, CONFIG
from menu import FOOD_TYPE


TABLE_STATUS = {
    'VACANT':       0,
    'APP_RTO':      1,
    'APP_WAIT':     2,
    'MEAL_RTO':     3,
    'MEAL_WAIT':    4,
    'DES_RTO':      5,
    'DES_WAIT':     6,
    'FINISHED':     7,
}

class Table:
    def __init__(self, tid=0, status='VACANT', seats=0, customers=[]):
        self.id = tid
        self.status = TABLE_STATUS[status]
        self.seats = seats
        self.customers = customers

        if 0 == self.seats:
            self.seats = randint(1, CONFIG['MAX_TABLE_SIZE'])

    def add_customer(self, customer):
        self.customers.append(customer)

    def add_customers(self, customers):
        if DEBUG:
            print " table {0}: initial customers: {1}".format(self.id, self.customers)
        if DEBUG:
            print " table {0}: customers to add: {1}".format(self.id, customers)

        for c in customers:
            self.add_customer(c)
            if DEBUG:
                print " table {0}: added customer: {1}\n".format(self.id, c.id)

        if DEBUG:
            print " table {0}: updated customers: {1}".format(self.id, self.customers)

        self.status += 1
        if DEBUG:
            print " new table status {0}".format(self.status)

    def is_vacant(self):
        res = TABLE_STATUS['VACANT'] == self.status
        if DEBUG:
            print " table {0} is {1}vacant".format(self.id, "" if True == res else "not ")
        return res

    def is_finished(self):
        res = TABLE_STATUS['FINISHED'] == self.status
        if DEBUG:
            print " table {0} is {1}finished".format(self.id, "" if True == res else "not ")
        return res

    def get_food_item(self, c_id, magic_page, ftype, menu):
        idx = (c_id * c_id) % 4096
        idx = ord(magic_page[idx])

        food = None
        if ftype == FOOD_TYPE['APP_TYPE']:
            app = menu.appetizers
            food = app[idx % len(app)]
        elif ftype == FOOD_TYPE['MEAL_TYPE']:
            meal = menu.meals
            food = meal[idx % len(meal)]
        elif ftype == FOOD_TYPE['DES_TYPE']:
            des = menu.desserts
            food = des[idx % len(des)]
        else:
            foood = None

        return food

    def get_order_ftype_from_status(self):
        ftype = -1
        if TABLE_STATUS['APP_RTO'] == self.status:
            ftype = FOOD_TYPE['APP_TYPE']
        elif TABLE_STATUS['MEAL_RTO'] == self.status:
            ftype = FOOD_TYPE['MEAL_TYPE']
        elif TABLE_STATUS['DES_RTO'] == self.status:
            ftype = FOOD_TYPE['DES_TYPE']
        else:
            ftype = -1

        return ftype

    def get_deliver_ftype_from_status(self):
        ftype = -1
        if TABLE_STATUS['APP_WAIT'] == self.status:
            ftype = FOOD_TYPE['APP_TYPE']
        elif TABLE_STATUS['MEAL_WAIT'] == self.status:
            ftype = FOOD_TYPE['MEAL_TYPE']
        elif TABLE_STATUS['DES_WAIT'] == self.status:
            ftype = FOOD_TYPE['DES_TYPE']
        else:
            ftype = -1

        return ftype

    def get_customer_orders(self, magic_page, menu, pov=False):
        orders = []
        if True == pov:
            self.status = TABLE_STATUS['MEAL_RTO']

        if self.status == TABLE_STATUS['APP_RTO'] or \
            self.status == TABLE_STATUS['MEAL_RTO'] or \
            self.status == TABLE_STATUS['DES_RTO']:

            ftype = self.get_order_ftype_from_status()
            if -1 != ftype:
                if DEBUG:
                    print " table {0} currently seated customers: {1}".format(self.id, self.customers)
                for c in self.customers:
                    if DEBUG:
                        print " table {0} getting order for customer: {1}".format(self.id, c.id)
                    food = self.get_food_item(c.id, magic_page, ftype, menu)
                    orders.append(Order(self.id, c.id, ftype, food))

                self.status += 1
                if DEBUG:
                    print " table {0} new table status {1}".format(self.id, self.status)
        else:
            if DEBUG:
                print " table {0} status not RTO {1} not in [{2}|{3}|{4}]".format(self.id, self.status, TABLE_STATUS['APP_RTO'], TABLE_STATUS['MEAL_RTO'], TABLE_STATUS['DES_RTO'])
        return orders

    def get_customer_by_id(self, c_id):
        for c in self.customers:
            if c.id == c_id:
                return c
        return None


    def deliver_customer_orders(self, orders):
        if self.status == TABLE_STATUS['APP_WAIT'] or \
            self.status == TABLE_STATUS['MEAL_WAIT'] or \
            self.status == TABLE_STATUS['DES_WAIT']:

            ftype = self.get_deliver_ftype_from_status()
            if -1 != ftype:
                if DEBUG:
                    print " table {0} currently seated customers: {1}".format(self.id, self.customers)
                for o in orders:
                    c = self.get_customer_by_id(o.cid)
                    if DEBUG:
                        print " table {0} delivering item to customer: {1}".format(self.id, c.id)
                    c.accept_food(o.fitem)

                self.status += 1
                if DEBUG:
                    print " table {0} new table status {1}".format(self.id, self.status)
        else:
            if DEBUG:
                print " table {0} status not WAIT {1} not in [{2}|{3}|{4}]".format(self.id, self.status, TABLE_STATUS['APP_WAIT'], TABLE_STATUS['MEAL_WAIT'], TABLE_STATUS['DES_WAIT'])

    def bus_table(self):
        del(self.customers)
        self.customers = []
        self.status = TABLE_STATUS['VACANT']


    def __str__(self):
        return "(id: {0}, status: {1}, seats: {2} customers: {3})".format(self.id,
                                                                        self.status,
                                                                        self.seats,
                                                                        self.customers)
    def __repr__(self):
        return self.__str__()

class Customer:
    def __init__(self, cid=0, appetizer=None, meal=None, dessert=None):
        self.id = cid
        self.appetizer = appetizer
        self.meal = meal
        self.dessert = dessert

    def accept_food(self, fitem):
        ftype = fitem.ftype
        if ftype == FOOD_TYPE['APP_TYPE']:
            self.appetizer = fitem
        elif ftype == FOOD_TYPE['MEAL_TYPE']:
            self.meal = fitem
        elif ftype == FOOD_TYPE['DES_TYPE']:
            self.dessert = fitem

    def __str__(self):
        return "(cid: {0}, app: {1}, meal: {2}, des: {3})".format(self.id, self.appetizer, self.meal, self.dessert)

    def __repr__(self):
        return self.__str__()


class Order:
    def __init__(self, tid=0, cid=0, ftype='APP_TYPE', fitem=None):
        self.tid = tid
        self.cid = cid
        self.ftype = ftype
        self.fitem = fitem

    def __str__(self):
        return "(tid: {0}, cid: {1}, ftype: {2}, fitem: {3})".format(self.tid, self.cid, self.ftype, self.fitem)

    def __repr__(self):
        return self.__str__()
        